parlai/core/worlds.py

```diff
@@ -975,6 +975,209 @@ def shutdown(self):
         self.world.shutdown()
 
 
+class DynamicBatchWorld(World):
+    def __init__(self, opt: Opt, world: Union[DialogPartnerWorld, MultiWorld]):
+        super().__init__(opt)
+        self.opt = opt
+
+        # agents is a placeholder just for super.reset
+        self.agents = []
+
+        # check some assumptions
+        if isinstance(
+            world, (ExecutableWorld, BatchWorld, HogwildWorld, MultiAgentDialogWorld)
+        ):
+            raise TypeError(
+                'World must be a DialogPartnerWorld or a '
+                'MultiWorld of DialogPartnerWorld'
+            )
+
+        if len(world.get_agents()) != 2:
+            raise AssertionError(
+                "Dynamic batch only works in a fixed dialog world with two agents."
+            )
+
+        if not hasattr(world.get_agents()[1], 'batch_act'):
```

Comment:
we have a function on worlds `get_task_agent`, should we have one for `get_model_agent` as well? rather than accessing `get_agents()[1]` directly

Reply:
Okay, I like that better.

Reply:
yes, i find this confusing in worlds everywhere where we have `agents[0]` and `agents[1]` and we are just supposed to know which is which